### PR TITLE
fix(cli,gateway): fully preserve DeepSeek reasoning_content across tool calls

### LINKED ISSUE #9501: DeepSeek reasoning_content error in multi-turn tool call conversations
## Issue Summary

DeepSeek reasoning models return HTTP 400 errors with the message "The reasoning_content in the thinking mode must be passed back to the API" during multi-turn conversations involving tool calls.

## Root Cause Analysis

Through extensive investigation, we identified four interconnected root causes that must be fixed together:

### 1. Missing extractReasoning Flag in Gateway
**Location**: `packages/kilo-gateway/src/provider.ts`
**Issue**: The Vercel AI SDK's `createOpenAI()` and `createOpenAICompatible()` calls were missing `extractReasoning: true`, preventing the SDK from creating structured reasoning content parts.

### 2. Hardcoded SDK Key in Transform Logic
**Location**: `packages/opencode/src/provider/transform.ts`
**Issue**: The reasoning content extraction logic hardcoded `"openaiCompatible"` as the providerOptions key, but OpenRouter models expect `"openrouter"`.

### 3. Missing Fallback for Historical Messages
**Location**: `packages/opencode/src/provider/transform.ts`
**Issue**: Assistant messages without reasoning parts (historical messages, tool-only responses) never get `reasoning_content` injected, causing API failures in subsequent requests.

### 4. Interleaved Not Auto-Enabled for Reasoning Models
**Location**: `packages/opencode/src/provider/provider.ts`
**Issue**: Models with `reasoning: true` don't automatically get interleaved capability enabled, preventing the reasoning extraction pipeline from triggering.

### Upstream Reference
- **Upstream PR**: anomalyco/opencode#24250 (`fix/deepseek-reasoning-complete`)
- **Commit**: `b6a7cdb` by user `knefenk`
- **Status**: Open, addresses the complete fix

### Previous Fix Attempts
- **PR #9470**: Partial fix cherry-picked from upstream #24146, only addressed Layer 2.3.1 (dynamic SDK key)
- **Result**: Incomplete - still missing gateway extractReasoning, unconditional fallback, and auto-interleaved
- **PR #9483**: Attempted gateway extractReasoning fix but closed prematurely

### Provider Path Analysis
The issue affects all DeepSeek provider paths:
- **Native DeepSeek API**: Uses `@ai-sdk/openai-compatible`
- **OpenRouter**: Uses `@openrouter/ai-sdk-provider`
- **Custom OpenAI-compatible**: Various providers pointing to DeepSeek endpoints

### SDK Key Mapping
The existing `sdkKey()` function correctly maps:
- `@openrouter/ai-sdk-provider` → `"openrouter"`
- `@ai-sdk/openai-compatible` → `undefined` (falls to default)
- `@kilocode/kilo-gateway` → `"openrouter"`

### Layer 1: Gateway extractReasoning
Add `extractReasoning: true` to both `createOpenAI()` and `createOpenAICompatible()` calls to enable structured reasoning content extraction.

### Layer 2: Transform Logic Updates
- **2.3.1**: Use `sdkKey()` for dynamic providerOptions key selection in interleaved block
- **2.3.2**: Add unconditional fallback block injecting empty `reasoning_content` for all assistant messages on reasoning models

### Layer 3: Auto-Enable Interleaved
Modify capability assembly to auto-enable `interleaved: { field: "reasoning_content" }` for any model with `reasoning: true`.

### Layer 4: Comprehensive Testing
Add tests covering all provider paths and edge cases.

## Impact Assessment

**Affected Scenarios**:
- Multi-turn conversations with DeepSeek reasoning models
- Tool call responses followed by continuation messages
- Sessions with historical messages (pre-fix)
- Both OpenRouter and native DeepSeek providers

**Expected Behavior After Fix**:
- No more 400 errors on reasoning_content
- Seamless multi-turn tool call conversations
- Backward compatibility with existing sessions
- Consistent behavior across all provider paths

### Manual Verification
1. OpenRouter + DeepSeek R1, multi-turn tool call → no 400 error
2. Custom OpenAI-compatible + DeepSeek API → same success
3. Load pre-fix session → no 400 on first new message
4. Multi-turn without tool calls, thinking mode → no error
5. Non-reasoning model (GPT-4) → no behavior change

### Automated Tests Added
- OpenRouter dynamic SDK key usage
- Historical message empty reasoning_content injection
- Auto-interleaved enablement
- Non-reasoning model regression protection

## Related Issues
- Closes #9471, #9482
- Supersedes previous partial fixes in #9470, #9483
- Upstream: anomalyco/opencode#24250

## Priority
High - Breaking functionality for DeepSeek reasoning model users in production scenarios.

### PR DESCRIPTION
**Pull request by scottnzuk**
## Context

The `reasoning_content` fix from PR #9470 was incomplete. Three additional layers needed: gateway extractReasoning, dynamic SDK key for OpenRouter, and unconditional injection for historical messages. This PR applies the complete fix matching upstream anomalyco/opencode#24250.

Fixes #9501 - DeepSeek reasoning_content error in multi-turn tool call conversations.

## Implementation

**Layer 1: Gateway extractReasoning**
- Added `extractReasoning: true` to gateway's `createOpenAI()` and `createOpenAICompatible()` calls
- Enables Vercel AI SDK to create structured reasoning content parts

**Layer 2: Transform Logic Updates**
- Used existing `sdkKey()` for dynamic providerOptions key selection in interleaved block
- Added unconditional fallback block injecting empty `reasoning_content` for all assistant messages on reasoning models
- Handles historical messages saved before the fix and tool-only responses

**Layer 3: Auto-Enable Interleaved**
- Modified capability assembly in `provider.ts` to auto-enable `interleaved` for reasoning models
- Ensures ANY model with `reasoning: true` gets the interleaved treatment

**Layer 4: Comprehensive Testing**
- Added 4 new tests covering OpenRouter dynamic key, historical messages, auto-interleaved, and regression protection
- All existing tests still pass (153/153)

## Screenshots

N/A (no UI changes)

## How to Test

Use a DeepSeek model with thinking mode, trigger a tool call (e.g., ask it to run a bash command), then send a follow-up message. Both turns should succeed without 400 errors.

Verify with both OpenRouter and native DeepSeek provider paths.

## Files Changed

- `packages/kilo-gateway/src/provider.ts` - Added extractReasoning flags
- `packages/opencode/src/provider/transform.ts` - Dynamic SDK key + unconditional fallback
- `packages/opencode/src/provider/provider.ts` - Auto-enable interleaved for reasoning models
- `packages/opencode/test/provider/transform.test.ts` - Added comprehensive tests
- `.changeset/deepseek-reasoning-complete.md` - Release notes

## Testing Results

- ✅ 153/153 tests pass in transform.test.ts
- ✅ `bun turbo typecheck` passes across all packages
- ✅ No spurious kilocode_change markers
- ✅ No opencode annotation issues

## Get in Touch

Questions about the implementation or testing scenarios.

### CODE REVIEW

**Comment by scottnzuk** [comment]
## ✅ Code Review Issues Addressed

I've fixed both issues identified by the code review:

### Issue 1: Auto-enabling interleaved for all reasoning models
**Fixed**: Made the auto-enablement specific to DeepSeek models only by checking  instead of applying to all reasoning models. This prevents incorrect  handling for providers that don't use this field format.

### Issue 2: Early return skipping string content handling  
**Fixed**: Removed the early return in the interleaved block and modified it to handle both array and string content messages. Now assistant messages with string content (not just array content) get proper  injection when they don't already have it.

### Additional Changes:
- Added comprehensive test for string content message handling
- Fixed TypeScript errors in the auto-enablement logic
- All tests pass (154/154) and type checking succeeds

The fixes ensure that only DeepSeek models get the interleaved auto-enablement, and all assistant message types (array and string content) are properly handled for reasoning content injection.

**Comment by scottnzuk** [comment]
## ✅ Code Review Issues Addressed

I've fixed both issues identified by the code review:

### Issue 1: Auto-enabling interleaved for all reasoning models
**Fixed**: Made the auto-enablement specific to DeepSeek models only by checking `model.id.includes('deepseek')` instead of applying to all reasoning models. This prevents incorrect `reasoning_content` handling for providers that don't use this field format.

### Issue 2: Early return skipping string content handling  
**Fixed**: Removed the early return in the interleaved block and modified it to handle both array and string content messages. Now assistant messages with string content (not just array content) get proper `reasoning_content` injection when they don't already have it.

### Additional Changes:
- Added comprehensive test for string content message handling
- Fixed TypeScript errors in the auto-enablement logic
- All tests pass (154/154) and type checking succeeds

The fixes ensure that only DeepSeek models get the interleaved auto-enablement, and all assistant message types (array and string content) are properly handled for reasoning content injection.

**Comment by scottnzuk** [comment]
## 📊 Updated Testing Results

Updated the testing results in the PR description:

- ✅ **154/154 tests pass** in transform.test.ts (was 153/153)
- ✅ • Packages in scope: @kilocode/cli, @kilocode/kilo-docs, @kilocode/kilo-gateway, @kilocode/kilo-i18n, @kilocode/kilo-jetbrains, @kilocode/kilo-telemetry, @kilocode/kilo-ui, @kilocode/plugin, @kilocode/sdk, @opencode-ai/app, @opencode-ai/desktop, @opencode-ai/desktop-electron, @opencode-ai/script, @opencode-ai/shared, @opencode-ai/storybook, @opencode-ai/ui, kilo-code
• Running typecheck in 17 packages
• Remote caching disabled
@kilocode/kilo-i18n:typecheck: cache hit, replaying logs db03c9f323690b0c
@opencode-ai/shared:typecheck: cache hit, replaying logs fc2b76b2d5b75a34
@kilocode/sdk:typecheck: cache hit, replaying logs 2f4d466616f4c6a1
@kilocode/kilo-i18n:typecheck: $ tsgo --noEmit
@opencode-ai/shared:typecheck: $ tsgo --noEmit
@kilocode/sdk:typecheck: $ tsgo --noEmit
@kilocode/plugin:typecheck: cache hit, replaying logs 338fe6f83792cbfd
@kilocode/plugin:typecheck: $ tsgo --noEmit
@opencode-ai/ui:typecheck: cache hit, replaying logs 8fd867f86f2e6608
@opencode-ai/ui:typecheck: $ tsgo --noEmit
@kilocode/kilo-gateway:typecheck: cache hit, replaying logs 2e57b4b2fda8189a
@kilocode/kilo-gateway:typecheck: $ tsgo --noEmit
@opencode-ai/app:typecheck: cache miss, executing cff9fc37e8c3b695
kilo-code:typecheck: cache miss, executing 5fe1aab7af98f966
@kilocode/kilo-telemetry:typecheck: cache hit, replaying logs ffcddaca397e3cc5
@kilocode/kilo-telemetry:typecheck: $ tsgo --noEmit
@kilocode/cli:typecheck: cache miss, executing b7e7d2e218ead293
@opencode-ai/app:typecheck: $ tsgo -b
@kilocode/cli:typecheck: $ tsgo --noEmit
kilo-code:typecheck: $ bun run check-types:extension && bun run check-types:webview
kilo-code:typecheck: $ bun script/typecheck.ts
@opencode-ai/desktop:typecheck: cache miss, executing d3aff5a1b72466f2
@opencode-ai/desktop-electron:typecheck: cache miss, executing e282aa0d53c73485
@opencode-ai/desktop:typecheck: $ tsgo -b
@opencode-ai/desktop-electron:typecheck: $ tsgo -b
kilo-code:typecheck: $ bun script/typecheck.ts --project webview-ui/tsconfig.json

 Tasks:    12 successful, 12 total
Cached:    7 cached, 12 total
  Time:    11.877s  passes across all packages  
- ✅ No spurious kilocode_change markers
- ✅ No opencode annotation issues

The additional test covers string content message handling that was added to address the code review feedback.

File: packages/opencode/src/provider/provider.ts [hunk 1]
```diff
@@ -1012,7 +1012,7 @@ function fromModelsDevModel(provider: ModelsDev.Provider, model: ModelsDev.Model
         video: model.modalities?.output?.includes("video") ?? false,
         pdf: model.modalities?.output?.includes("pdf") ?? false,
       },
-      interleaved: model.interleaved ?? false,
+      interleaved: model.interleaved ?? (model.reasoning ? { field: "reasoning_content" } : false), // kilocode_change
```
[the remark this answers is not in the document]
Comment (scottnzuk) [reply packages/opencode/src/provider/provider.ts]: Code Review Issues Addressed
I've fixed both issues identified by the code review:

Issue 1: Auto-enabling interleaved for all reasoning models
Fixed: Made the auto-enablement specific to DeepSeek models only by checking instead of applying to all reasoning models. This prevents incorrect handling for providers that don't use this field format.

Issue 2: Early return skipping string content handling
Fixed: Removed the early return in the interleaved block and modified it to handle both array and string content messages. Now assistant messages with string content (not just array content) get proper injection when they don't already have it.

Additional Changes:
Added comprehensive test for string content message handling
Fixed TypeScript errors in the auto-enablement logic
All tests pass (154/154) and type checking succeeds
The fixes ensure that only DeepSeek models get the interleaved auto-enablement, and all assistant message types (array and string content) are properly handled for reasoning content injection.

File: packages/opencode/src/provider/transform.ts [hunk 2]
```diff
@@ -185,6 +185,7 @@ function normalizeMessages(
 
   if (typeof model.capabilities.interleaved === "object" && model.capabilities.interleaved.field) {
     const field = model.capabilities.interleaved.field
+    const sdk = sdkKey(model.api.npm) ?? "openaiCompatible"
     return msgs.map((msg) => {
```
[the remark this answers is not in the document]
Comment (scottnzuk) [reply packages/opencode/src/provider/transform.ts]: Code Review Issues Addressed
I've fixed both issues identified by the code review:

Issue 1: Auto-enabling interleaved for all reasoning models
Fixed: Made the auto-enablement specific to DeepSeek models only by checking model.id.includes('deepseek') instead of applying to all reasoning models. This prevents incorrect reasoning_content handling for providers that don't use this field format.

Issue 2: Early return skipping string content handling
Fixed: Removed the early return in the interleaved block and modified it to handle both array and string content messages. Now assistant messages with string content (not just array content) get proper reasoning_content injection when they don't already have it.

Additional Changes:
Added comprehensive test for string content message handling
Fixed TypeScript errors in the auto-enablement logic
All tests pass (154/154) and type checking succeeds
The fixes ensure that only DeepSeek models get the interleaved auto-enablement, and all assistant message types (array and string content) are properly handled for reasoning content injection.

**Comment by scottnzuk** [comment]
@kilocode-bot anything else needed from me?

**Comment by scottnzuk** [comment]
Well after 3 revisions, endless token use and testing, still can't get to the bottom of this.. if anyone else wants to review this and see if I'm on the right track, and/or fix it, that would be amazing.

For now, my fix is: when the error shows up, swap from Deepseek-V4 Flash to Deepseek-Chat. Oddly enough, they are actually both the same DeepSeek pass and API calls to chat to the new flash. But it allows me to skip over the error and carry on.

ugh lol

**Comment by edulml98** [comment]
Issue is related to openrouter https://github.com/OpenRouterTeam/ai-sdk-provider/pull/487
edit: bump the ai-sdk-provider to the latest version and test again

**Comment by scottnzuk** [comment]
> Issue is related to openrouter [OpenRouterTeam/ai-sdk-provider#487](https://github.com/OpenRouterTeam/ai-sdk-provider/pull/487) edit: bump the ai-sdk-provider to the latest version and test again

that would be correct for kilo version but i use direct api that or is that using open router back end as well?

**Comment by babang1da** [comment]
please fix it

**Comment by TempUser13** [comment]
OpenRouter's AI-SDK-Provider is now on v2.8.1 https://github.com/OpenRouterTeam/ai-sdk-provider/releases/tag/2.8.1 which fixes the reasoning content array issue https://github.com/OpenRouterTeam/ai-sdk-provider/pull/487. As Deepseek has cut the API costs for V4 Pro by 75% until May 5th, would you be able to bump the version in KiloCode as soon as possible please? Thank you for the work that you do.

**Comment by lambertjosh** [comment]
The SDK has been bumped as of 7.2.25. https://github.com/Kilo-Org/kilocode/pull/9526. Closing this, as it is working. If not please open a separate issue.

**Comment by lambertjosh** [comment]
Thanks for the contribution and collaboration here, we were on parallel tracks I think.
